### DISCUSSION TITLE
[nuxt] A composable that requires access to the Nuxt instance was called outside of a plugin

### DISCUSSION THREAD

**Discussion by cdwmhcc** [nuxt/nuxt#23992] [Questions] [answered]
### Environment

------------------------------
- Operating System: Windows_NT
- Node Version:     v20.9.0
- Nuxt Version:     3.8.0
- CLI Version:      3.9.1
- Nitro Version:    2.7.1
- Package Manager:  pnpm@8.9.2
- Builder:          -
- User Config:      devtools, modules
- Runtime Modules:  @pinia/nuxt@0.5.1
- Build Modules:    -
------------------------------

### Reproduction

https://stackblitz.com/edit/nuxt-starter-m1wqdm?file=pages%2Findex.vue

Need to click the button to see the actual error.
![image](https://github.com/nuxt/nuxt/assets/8030134/8e545e28-7a85-4081-9268-27114c3a8e56)

### Describe the bug

```
[nuxt] A composable that requires access to the Nuxt instance was called outside of a plugin, Nuxt hook, Nuxt middleware, or Vue setup function. This is probably not a Nuxt bug. Find out more at `https://nuxt.com/docs/guide/concepts/auto-imports#using-vue-and-nuxt-composables`.
```

I looked through the document but don't quite understand it.
https://nuxt.com/docs/api/composables/use-nuxt-app#runwithcontext

**Comment by bolt-new-by-stackblitz** [comment]
<a href='https://stackblitz.com/~/github.com/nuxt/nuxt/issues/23991?repros=nuxt-starter-m1wqdm'><img src='https://developer.stackblitz.com/img/fix_this_issue_small.svg' alt='Fix this issue in StackBlitz Codeflow' align='left' width='117' height='20'></a> _Start a new pull request in [StackBlitz Codeflow](https://stackblitz.com/~/github.com/nuxt/nuxt/issues/23991?repros=nuxt-starter-m1wqdm)._

- [**nuxt-starter-m1wqdm** (Open reproduction)](https://stackblitz.com/edit/nuxt-starter-m1wqdm?issueRepo=nuxt/nuxt&issueNumber=23991)

**Comment by cdwmhcc** [comment]
@danielroe what happened?

**Reply by huang-julien** [reply]
Hello, we use a transform plugin to keep the context sync. However this can't be applied to nested async/await; Meaning you have to wrap what uses the nuxtContext within `runWithContext`

````ts
  if (refresh_cookie.value) {
    const { data } = await useFetch<AuthenticationData>('/api/auth/login');
    if (data.value?.refresh_token) {
      nuxtApp.runWithContext(() => {
        useCookie('refresh_token', {
          maxAge: data.value?.expires ?? 0,
        });
      })
      
      refresh_cookie.value = data.value?.refresh_token;
      appStore.access_token = data.value?.access_token!;
      appStore.authenticated = true;
     
      await nuxtApp.runWithContext(() => useReadMe()) 
    }
  }
```

**Reply by cdwmhcc** [reply, marked as answer]
`Cannot stringify arbitrary non-POJOs`
```ts
            if (expires <= 2100000000) {
              nuxtApp.runWithContext(() => {
                this.refreshTimeout = setTimeout(() => this.refresh(), expires - 10000)
              })
            }
```

**Comment by xuhao008** [comment]
请问解决了吗，我也遇到了这个问题

**Reply by cdwmhcc** [reply]
```
await nuxtApp.runWithContext(() => useReadMe()) 
```

For await and some methods that need to be nested, just use it to wrap them.
